### ISSUE TITLE
Delay before showing insertion caret after 'o' command

### ISSUE THREAD

**Issue by ricksladkey** [jaredpar/VsVim#2424]
For the past few months, mostly using the head of VsVim master or related branches, I have noticed a significant delay after typing the `o` (open new line) command before the insertion caret is displayed. The delay is a least one second. My experience has mostly been with Visual Studio 2017.

Now that VsVim 2.6.0 has been released, I've tested 2.6.0 against Visual Studio 2015 and it does not have the same problem. There isn't anything obvious that VsVim could be doing to cause the delay and using the simple `i` or `a` command only sometimes causes the delay with VS2017.

Trying to understand the constellation of symptoms, I am sometimes able to induce the delay using `<C-S-F12>` from normal mode. When it happens the Visual Studio insertion caret takes over a second to appear even though VsVim (for the most part) is entirely out of the loop. As a result, I am suspecting some change in Visual Studio 2017 itself is the root cause of the problem.

Anyway, I was wondering if anyone else can reproduce this delay. It's just irritating enough to be distracting!

**Comment by jaredpar** [comment]
I haven't seen this yet. Nor did I see it on any of the internal builds of VS2017.

Have you tried profiling and seeing where the delay is? Given the symptoms it should show up pretty clear on a profile. If you can grab a perfview trace I can take a look at it.

**Comment by riandrake** [comment]
I am also experiencing this issue, only started happening (noticeably) recently:

![repro](https://user-images.githubusercontent.com/5866058/48965109-d8fc5980-f01b-11e8-8a52-4269a54f8d1b.gif)

**Comment by ricksladkey** [comment]
I came up with an experiment to see if VsVim was the source of the problem. Some context is the fact that VsVim completely constructs, manages and displays the caret when in normal mode but relies on Visual Studio to display the caret when in insert mode. However, in exclusive select mode, VsVim supports a simulated caret this is substantially indistinguishable from the Visual Studio caret. By using the exclusive selection caret instead of the insertion caret, we can force VsVim to completely manage the caret in both normal and insert modes.

So the experiment makes this very simple change to the block caret controller:

![image](https://user-images.githubusercontent.com/2680524/48971553-024fc200-efe9-11e8-95a2-8d28cf82c9ba.png)

In testing, the insertion caret now appears reliably and instantly. Since this completely fixes the problem, the question is then:

- Is there any good reason to let Visual Studio manage the caret in insert mode?

If not, then we don't have to worry about why Visual Studio is not displaying the caret, as it only affects how quickly the caret will be display when we disable VsVim.

**Comment by jaredpar** [comment]
Is this just a UI issue then and not an actual performance problem? By that I mean in the cases where the cursor display is delayed can typing begin immediately? 

I'm not opposed to just having VsVim control the caret in all cases. Originally I didn't because cursor drawing was still quite new to me and I was interested in minimizing the impact of my change. 

But this will cause a problem. Visual Assist uses whether or not the cursor is visible to cue whether they should be displaying intellisense on key presses. Before this change they displayed intellisense on every alpha key press. Hence using <kbd>j</kbd> in normal mode triggered intellisense for 'j' even though it hadn't actually been typed. There devs reached out (super nice BTW) and eventually worked out that they should effectively disable intellisense whenever the cursor was hidden. If we completely control the cursor in VsVim it seems like that will break Visual Assist.

**Comment by ricksladkey** [comment]
It didn't occur to me to report whether the editor is functional or "hung", but yes, it is responsive to typing immediately. This implies the restoration of the caret is occurring on some sort of delay timer within the Visual Studio editor.

My experiment strongly supports that the setting of the `IsHidden` boolean does occur instantly upon switching to insert mode. All this points to a Visual Studio issue rather than a VsVim issue. If there were no workaround, we could try to get the issue resolved through Visual Studio channels, but that approach has "when" and "if" risks.

At this point, it is bothersome enough to me that I support using my workaround  unless Visual Assist is active. Then, if some user both uses Visual Assist and experiences the delay, then they can lead the effort of getting Visual Studio to recognize and fix the problem.

**Comment by jaredpar** [comment]
> At this point, it is bothersome enough to me that I support using my workaround unless Visual Assist is active

That seems reasonable to me. 

I wonder if it's just a simple timing issue. There are two caret timers in this scenario: VsVim and the OS / VS one. Possibly on your machine you're just hitting the longest possible delay between the timers.

**Comment by ricksladkey** [comment]
@jaredpar The universal constant in blinking caret management is that if the caret goes from not being displayed to being displayed, or if the caret changes position, then a new full "on" cycle should start immediately. VsVim, Visual Studio and Windows all (normally) behave this way.

You can see for yourself in @riandrake's first comment above that it looks like at least the first complete "on" cycle is being skipped:

Desired:

```
+---+   +---+
|   |   |   |
+   +---+   +---
```

Actual:

```
        +---+
        |   |
+---+---+   +---
```

To put it another way, if it were simply a timer synchronization issue, and "on"/"off" cycles are 1/2 second, then the caret should become visible within half a second. I'd say it is either instant (the correct behavior) or over a second (the reported issue).

**Comment by jaredpar** [comment]
@ricksladkey 

> The universal constant in blinking caret management is that if the caret goes from not being displayed to being displayed, or if the caret changes position, then a new full "on" cycle should start immediately.

That's what my intuition was on how it should behave but didn't know enough to assert that was the behavior. Sounds like this is definitely a bug of some kind then in Visual Studio.
